### DISCUSSION TITLE
pretty tree output (like pest.rs sandbox)

### DISCUSSION THREAD

**Discussion by shanemikel** [pest-parser/pest#650] [General]
I can't seem to find any methods or functions for the pretty tree output you have in the pest.rs sandbox tool.

The output is like so:
```
- RuleWithMultipleChildren
  - RuleWithOneChild > RuleWithMultipleChildren
    - RuleWithOneChild > TerminalRule: "tok1"
    - RuleWithOneChild > TerminalRule: "tok2"
```

I find that this has been really nice for debugging, so I implemented my own (pretty ugly) version of it here (https://gist.github.com/shanemikel/72e4a71452a7663749c6cb030136b855).  From what I can tell, it produces the exact same output as the website.

I would very much appreciate if the already implemented version would be exposed in the pest library itself.

---

This brings me to a question.  The following line from this function (line no. 23... can't figure out how to link a line from the gist):
```
let children: Vec<_> = pairs.clone().into_iter().collect();
let child_count = children.len();
```
Now does this simply clone the iterator, or make a deep copy of the entire parse tree?

**Comment by dragostis** [comment]
The code that formats the output is in the [site repo](https://github.com/pest-parser/site/blob/master/src/main.rs#L70).

While this might seem like a good idea, it will probably not be an area of focus for the next major version of pest since we're exploring ways of deprecating the `Pair`/`Pairs` API in favor of a simpler, more flexible API.

**Comment by shanemikel** [comment]
@dragostis Thank god (or just the Pest devs)!  The Pairs API is awful to use.
**EDIT** Sorry author of Pairs...

**Comment by shanemikel** [comment]
Holy crap that implementation is soo much cleaner than mine.
